### DISCUSSION TITLE
New release of QCheck.0.18.1 ?

### DISCUSSION THREAD

**Discussion by vch9** [c-cube/qcheck#211] [General]
We are introducing QCheck2 tests in our code base (Tezos). One developer got the stack overflow fixed by @jmid in https://github.com/c-cube/qcheck/pull/160.

The `CHANGELOG` does not seem to have been updated in a while. From a quick look I see:
* https://github.com/c-cube/qcheck/pull/181
* https://github.com/c-cube/qcheck/pull/183
* https://github.com/c-cube/qcheck/pull/160

which impact users.

What do you think about a new release @c-cube @jmid, I believe this should be a fix-only release

**Comment by c-cube** [comment]
You mean a release at f750ccf623e223b50aed60f791fef73f71b45fc9 , just after #160 is merged?

A lot seems to have happened on master, I think releasing current master would deserve 0.19 at least.

**Reply by vch9** [reply]
A "fix" release after https://github.com/c-cube/qcheck/commit/f750ccf623e223b50aed60f791fef73f71b45fc9 could solve my issue yes. But, yes we could refine the `CHANGELOG` to see if a `0.19.0` is worth.

**Reply by c-cube** [reply]
typically I tend to write the changelog when I release. But yeah, look at `gitg -a` if it's installed on  your machine :)

**Reply by c-cube** [reply]
@jmid do you agree with me making a small fix release at f750ccf ? we can also discuss 0.19, if it's not too early for that.

**Reply by jmid** [reply]
A 0.18.1 bugfix release sounds good 👍‍
I think it would make sense to include #183 aka 906af37f1bf47273f5c3d90435187c543d0d99d7 fixing bugs reported in #180 and #198
to avoid releasing something that contains other known bugs.
Some time went by before it was reviewed and merged (I added a few tests while doing so).

Other than those, there's the `t2`-`t9` tuple generators and a bunch of typo fixes.
The rest I think are part of `ppx_deriving_qcheck` (which will be released separately IIUC).

**Reply by c-cube** [reply]
so picking 906af37 would cover everything? I can do that.

edit: there's actually a *ton* of stuff that happened in the mean time. Most of it seems to be tests, but still, makes me uncomfortable to make a bugfix release with so many changes.

Should we cherry-pick instead? Or just go for 0.19 ?

**Reply by jmid** [reply]
Unfortunately the history around 906af37 is not linear: other things were developing and being fixed in parallel.

3ee3c01 where CI got back up and running might be a better choice (green lights for a release are a good sign... 😀).

**Reply by vch9** [reply]
I think having a `0.18.1` on https://github.com/c-cube/qcheck/commit/3ee3c0110a55a31e6416b0158dfc8f5e47d6ec23 it is fine as it does not break anything already existing (no deprecation AFAIK)

**Reply by vch9** [reply]
Reviving this discussion as we currently need the fix from @jmid in https://github.com/c-cube/qcheck/commit/f750ccf623e223b50aed60f791fef73f71b45fc9 in our project. I'm not sure I can help you with that matter

**Reply by c-cube** [reply]
I think, last time I looked, the earliest commit that would work also had a lot of new features under it, which imho precludes a point release?

Why don't we go to 0.19 now, anyway? There's a lot of new features already.

**Reply by vch9** [reply]
> I think, last time I looked, the earliest commit that would work also had a lot of new features under it, which imho precludes a point release?

Don't you think we could cherry-pick the commit itself?

> Why don't we go to 0.19 now, anyway? There's a lot of new features already

But yes, 0.19 is greater than a fix release but does require slightly more actions.

**Reply by c-cube** [reply]
I think the cherry-pick didn't work, but if you make a branch (based off 0.18) that does it and passes tests, then why not :)

**Reply by jmid** [reply]
I'm surprised you've chosen to go with QCheck2 already @vch9! I'm still hoping to find time to improve the Rose tree shrinkers...

I'm fine with either a 0.18.1 or a 0.19.0 👍‍
As I mentioned, I think the former should include #183 to avoid a bugfix release with other known bugs.
I wasn't considering cherry-picking at the time and therefore suggested 3ee3c01 as a reasonable compromise.

A 0.19.0 release probably involves a bit more work as @vch9 remarks: For example, settling on a revised design in #223 first would be preferable to breaking the QCheck2.Gen interface after both a 0.18.0 and a 0.19.0 release.

How do we plan to release qcheck{,-core,-ounit,-alsotest} and ppx_deriving_qcheck?
- Independently when each has something to share with the world?
- Simultaneously coordinated releases as separate opam packages?
- Version-aligned - so that ppx_deriving_qcheck jumps to 0.18.1/0.19.0 as the others?

From, e.g., #197 it seems a public release of ppx_deriving_qcheck would be welcome (I would certainly also welcome it!).
I can see #192 lists #209 as a requirement. Will this be the last release requirement @vch9 - or do you also plan to include #189?
Perhaps we should go for a combined ppx_deriving_qcheck-release +  QCheck bugfix release?

**Reply by vch9** [reply]
> I'm surprised you've chosen to go with QCheck2 already @vch9! I'm still hoping to find time to improve the Rose tree shrinkers...

So far it has helped us fixing bugs, we had to reduce some size generators for string and lists to prevent a much longer execution time.

> A 0.19.0 release probably involves a bit more work as @vch9 remarks: For example, settling on a revised design in https://github.com/c-cube/qcheck/pull/223 first would be preferable to breaking the QCheck2.Gen interface after both a 0.18.0 and a 0.19.0 release.

Having #223 would be better yes, I'm afraid I kind of lack time to work on this for the moment but I'll try to at least finished the naming consistency I started. They were some improvements about missing generators which I'm not sure I can dive into right now.

> How do we plan to release qcheck{,-core,-ounit,-alsotest} and ppx_deriving_qcheck?
>
>    * Independently when each has something to share with the world?
>    * Simultaneously coordinated releases as separate opam packages?

Apart from `ppx_deriving_qcheck` I think the three should be upgraded to 0.19 together, just like @c-cube did in https://github.com/ocaml/opam-repository/pull/19207.

> Version-aligned - so that ppx_deriving_qcheck jumps to 0.18.1/0.19.0 as the others?

The deriver should probably follows the `qcheck-core` version yes.

> From, e.g., https://github.com/c-cube/qcheck/issues/197 it seems a public release of ppx_deriving_qcheck would be welcome (I would certainly also welcome it!).
I can see https://github.com/c-cube/qcheck/issues/192 lists https://github.com/c-cube/qcheck/pull/209 as a requirement. Will this be the last release requirement @vch9 - or do you also plan to include https://github.com/c-cube/qcheck/issues/189?

#209 is the most important one I believe, #189 should be done at some point, one day.

> Perhaps we should go for a combined ppx_deriving_qcheck-release + QCheck bugfix release?

We could yes, but I also think we could treat the deriver and the bugfix release separately. The bugfix seems more important as people do not use the deriver yet. What I mean is the bugfix release should not be delayed for a bug fix release, but it could make sense for the 0.19.

**Reply by vch9** [reply]
I also think https://github.com/c-cube/qcheck/commit/3ee3c0110a55a31e6416b0158dfc8f5e47d6ec23 could be the 0.18.1

**Reply by c-cube** [reply]
but there's new features in between that commit and 0.18…

**Reply by vch9** [reply]
There is? I had a quick look at the commits names and except for deriver related merge request I did not see any new features.

I can try to as quickly as possible get #223 merged and release a 0.19. (If we go for a 0.18.1 this would be a bit delayed.)

**Reply by c-cube** [reply]
My bad, it seems ok. I misread a commit title about tests and thought it added a combinator :sweat: . I can try and kick off a release from there if you want?

**Reply by vch9** [reply]
That'd be awesome, ping me if you want me to double-check if there is no new features introduced.

**Reply by c-cube** [reply]
https://github.com/ocaml/opam-repository/pull/20636
